Support independent builds for cpp extension tests + apply to libtorch_agnostic tests

jbschlosser opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #153855
* #153274
* __->__ #153264

Related: #148920

This PR:
* Provides a helper `install_cpp_extension(extension_root)` for building C++ extensions. This is intended to be used in `TestMyCppExtension.setUpClass()`
    * Updates libtorch_agnostic tests to use this
* Deletes preexisting libtorch_agnostic tests from `test/test_cpp_extensions_aot.py`
    * Fixes `run_test.py` to actually run tests in `test/cpp_extensions/libtorch_agnostic_extension/test/test_libtorch_agnostic.py` to avoid losing coverage. This wasn't being run due to logic excluding tests that start with "cpp"; this is fixed now

After this PR, it is now possible to run:
```
python test/cpp_extensions/libtorch_agnostic_extension/test/test_libtorch_agnostic.py
```

and the test will build the `libtorch_agnostic` extension before running the tests.

jbschlosser committed:
Support independent builds for cpp extension tests + apply to libtorch_agnostic tests

[ghstack-poisoned]

jbschlosser committed:
Update on "Support independent builds for cpp extension tests + apply to libtorch_agnostic tests"




[ghstack-poisoned]

jbschlosser committed:
Update on "Support independent builds for cpp extension tests + apply to libtorch_agnostic tests"




[ghstack-poisoned]

janeyx99 commented on test/cpp_extensions/libtorch_agnostic_extension/test/test_libtorch_agnostic.py:
```diff
@@ -39,6 +49,8 @@ def test_slow_sgd(self, device):
 
     @onlyCUDA
     def test_identity_does_not_hog_memory(self, device):
+        import libtorch_agnostic
```
Why not put this in line 22?

jbschlosser replied:
Putting the import there doesn't import to global scope, so the test functions still won't be able to see it. There's a hacky way to avoid updating the test functions:

```python
globals()["libtorch_agnostic"] = importlib.import_module("libtorch_agnostic")
```

janeyx99 replied:
nawwwww what u have rn is more explicit i agree

janeyx99 commented on torch/testing/_internal/common_utils.py:
```diff
@@ -5617,6 +5617,36 @@ def remove_cpp_extensions_build_root():
             shutil.rmtree(default_build_root, ignore_errors=True)
 
+
+def install_cpp_extension(extension_root):
```
is this copied from run_test.py?

jbschlosser replied:
with some changes, yes :) eventually I plan to purge run_test.py of this logic entirely

janeyx99 reviewed (commented):
can you share a link to the CI shard running the libtorch agnostic tests in CI now?

thanks for fixing this--this brings me much joy

jbschlosser committed:
Update on "Support independent builds for cpp extension tests + apply to libtorch_agnostic tests"


Related: #148920

This PR:
* Provides a helper `install_cpp_extension(extension_root)` for building C++ extensions. This is intended to be used in `TestMyCppExtension.setUpClass()`
    * Updates libtorch_agnostic tests to use this
* Deletes preexisting libtorch_agnostic tests from `test/test_cpp_extensions_aot.py`
    * Fixes `run_test.py` to actually run tests in `test/cpp_extensions/libtorch_agnostic_extension/test/test_libtorch_agnostic.py` to avoid losing coverage. This wasn't being run due to logic excluding tests that start with "cpp"; this is fixed now

After this PR, it is now possible to run:
```
python test/cpp_extensions/libtorch_agnostic_extension/test/test_libtorch_agnostic.py
```

and the test will build the `libtorch_agnostic` extension before running the tests.

[ghstack-poisoned]

jbschlosser commented:
@pytorchbot merge

jbschlosser committed:
Update on "Support independent builds for cpp extension tests + apply to libtorch_agnostic tests"


Related: #148920

This PR:
* Provides a helper `install_cpp_extension(extension_root)` for building C++ extensions. This is intended to be used in `TestMyCppExtension.setUpClass()`
    * Updates libtorch_agnostic tests to use this
* Deletes preexisting libtorch_agnostic tests from `test/test_cpp_extensions_aot.py`
    * Fixes `run_test.py` to actually run tests in `test/cpp_extensions/libtorch_agnostic_extension/test/test_libtorch_agnostic.py` to avoid losing coverage. This wasn't being run due to logic excluding tests that start with "cpp"; this is fixed now

After this PR, it is now possible to run:
```
python test/cpp_extensions/libtorch_agnostic_extension/test/test_libtorch_agnostic.py
```

and the test will build the `libtorch_agnostic` extension before running the tests.

[ghstack-poisoned]

jbschlosser committed:
Update on "Support independent builds for cpp extension tests + apply to libtorch_agnostic tests"


Related: #148920

This PR:
* Provides a helper `install_cpp_extension(extension_root)` for building C++ extensions. This is intended to be used in `TestMyCppExtension.setUpClass()`
    * Updates libtorch_agnostic tests to use this
* Deletes preexisting libtorch_agnostic tests from `test/test_cpp_extensions_aot.py`
    * Fixes `run_test.py` to actually run tests in `test/cpp_extensions/libtorch_agnostic_extension/test/test_libtorch_agnostic.py` to avoid losing coverage. This wasn't being run due to logic excluding tests that start with "cpp"; this is fixed now

After this PR, it is now possible to run:
```
python test/cpp_extensions/libtorch_agnostic_extension/test/test_libtorch_agnostic.py
```

and the test will build the `libtorch_agnostic` extension before running the tests.

[ghstack-poisoned]

jbschlosser commented:
@janeyx99 i'm still getting a windows failure after shortening the filename. It looks like these tests may have been blocklisted for Windows in the past?

https://github.com/pytorch/pytorch/blob/3102ae6798607f380fddd03bb55df375e5ed044d/test/run_test.py#L1605-L1612

jbschlosser committed:
Update on "Support independent builds for cpp extension tests + apply to libtorch_agnostic tests"


Related: #148920

This PR:
* Provides a helper `install_cpp_extension(extension_root)` for building C++ extensions. This is intended to be used in `TestMyCppExtension.setUpClass()`
    * Updates libtorch_agnostic tests to use this
* Deletes preexisting libtorch_agnostic tests from `test/test_cpp_extensions_aot.py`
    * Fixes `run_test.py` to actually run tests in `test/cpp_extensions/libtorch_agnostic_extension/test/test_libtorch_agnostic.py` to avoid losing coverage. This wasn't being run due to logic excluding tests that start with "cpp"; this is fixed now

After this PR, it is now possible to run:
```
python test/cpp_extensions/libtorch_agnostic_extension/test/test_libtorch_agnostic.py
```

and the test will build the `libtorch_agnostic` extension before running the tests.

[ghstack-poisoned]
